external-crates/move/crates/move-package/src/compilation/compiled_package.rs

```diff
@@ -756,3 +765,123 @@ pub(crate) fn make_source_and_deps_for_compiler(
     };
     Ok((source_package_paths, deps_package_paths))
 }
+
+/// partitions `deps` by whether we need to compile dependent packages with a
+/// prior toolchain (which we find by looking at Move.lock contents) or
+/// whether we can compile them with the current binary.
+fn partition_deps_by_toolchain(
+    deps: Vec<PackagePaths>,
+) -> Result<(Vec<PackagePaths>, Vec<PackagePaths>)> {
+    let mut deps_for_current_compiler = vec![];
+    let mut deps_for_prior_compiler = vec![];
+    for dep in deps {
+        let a_source_path = dep.paths[0].as_str();
+        let root = SourcePackageLayout::try_find_root(Path::new(a_source_path))?;
+        let lock_file = root.join("Move.lock");
+        if !lock_file.exists() {
+            // Q: Behavior to pick when package has no lock file. Right now we choose the current compiler (we could use the "legacy" one instead).
+            deps_for_current_compiler.push(dep);
+            continue;
+        }
+
+        let mut lock_file = File::open(lock_file)?;
+        let toolchain_version = lock_file::schema::ToolchainVersion::read(&mut lock_file)?;
+        match toolchain_version {
+            None => {
+                // Q: Behavior to pick when package has no lock file. Right now we choose the current compiler (we could use the "legacy" one instead).
+                deps_for_current_compiler.push(dep)
+            }
+            Some(ToolchainVersion {
+                compiler_version, ..
+            }) if compiler_version == env!("CARGO_PKG_VERSION") => {
+                // This dependency requires the same compiler version we're on.
+                deps_for_current_compiler.push(dep)
+            }
+            Some(toolchain_version) => {
+                // Compile and mark that we compiled this dep with a prior compiler.
+                download_and_compile(root, toolchain_version)?;
+                deps_for_prior_compiler.push(dep)
+            }
+        }
+    }
+    Ok((deps_for_current_compiler, deps_for_prior_compiler))
+}
+
+fn download_and_compile(
+    root: PathBuf,
+    ToolchainVersion {
+        compiler_version,
+        edition,
+        flavor,
+    }: ToolchainVersion,
+) -> Result<()> {
+    let binaries_path = std::env::var("TOOLCHAIN_BINARIES")?; // E.g., ~/.move/binaries
+    let dest_dir = Path::new(binaries_path.as_str());
+    let dest_version = dest_dir.join(compiler_version.clone());
+    let platform = "macos-arm64"; // Hardcoded platform: reviewer please ignore this for now.
+    let dest_binary = dest_version.join(format!("target/release/sui-{}", platform));
+    let dest_binary_os = OsStr::new(dest_binary.as_path());
+
+    if !dest_binary.exists() {
+        // Download if binary does not exist.
```

Comment:
I'm just using `curl`/`tar` to test this end-to-end, it can and will probably be replaced with native Rust `reqwest`, etc.